### PR TITLE
Fix optional types in queries

### PR DESCRIPTION
**Pull request by standy**
Fix for #3638
Related #3463 #3060 #2590

This MR makes all nullable fields be optional (for typescript and flow), if `avoidOptionals` option is not true. 

There is an inconsistency between the top-level types and the query types. Example from `dev-test`: 
https://github.com/dotansimha/graphql-code-generator/blob/master/dev-test/githunt/types.preResolveTypes.ts
```ts
// This is the top-level type and description is optional
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

// This is query type, and description is mandatory now
export type CommentQuery = {
    repository: {
      __typename?: 'Repository';
      description: Maybe<string>;
      open_issues_count: Maybe<number>;
      stargazers_count: number;
      full_name: string;
      html_url: string;
    };
  }>;
};
```

Same bug for `preresolveTypes: false`
Example from https://github.com/dotansimha/graphql-code-generator/blob/master/dev-test/githunt/types.ts
```ts
// This is top-level type
export type Query = {
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};

// This is query type, notice currentUser become mandatory
export type CommentQuery = { __typename?: 'Query' } & {
  currentUser: Maybe<{ __typename?: 'User' } & Pick<User, 'login' | 'html_url'>>;
}
```

My fix is actually easy. There are two methods: `formatNamedField` and `wrapTypeWithModifiers`, they should be applied always together, and `formatNamedField` should add optional mark when it needed (line [53](https://github.com/dotansimha/graphql-code-generator/pull/3639/commits/ea69ed97042a34ac9c34bb65ad0baf2e63f99258#diff-b05502b152b19c27071a4c3bffb8d02fR54))

### CODE REVIEW

**Comment by dotansimha** [comment]
Hi @standy ! 

This seems like a big PR with more changes, can you please explain your changes there? 
because the bug effected `preResolveTypes` types only as your said, but it seems like all examples under `dev-test` are updated with changes, and you changed some more files? 

Thanks!

**Comment by standy** [comment]
Hello, @dotansimha 

I did update my PR. And I found a few similar issues, which is fixed here.

**Comment by dotansimha** [comment]
Just went through the changes in this PR. This is great!
Thank you so much @standy  :)

**Comment by david-golightly-leapyear** [comment]
🚨 I think this "fix" was incorrect. It seems to break the behavior described here: https://github.com/dotansimha/graphql-code-generator/issues/3060#issuecomment-570174939

We tried upgrading to 0.13.1 and noticed a lot of client-side generated query types were now suddenly "optional". A field that is present in the client-side query should never be "optional", although its value may be `null` if it is not marked non-nullable.

So this PR introduced incorrect behavior and IMO should be reverted. Happy to discuss further - I know a lot of work went into it! It's a subtle distinction, but it's important not to confuse "client" and "server" types in implementations - they may be different, but the code generation should account for this, and marking every "nullable" field as "optional" confuses the two.

**Comment by standy** [comment]
@david-golightly-leapyear 
Take a look into `avoidOptionals` option, I think this is what you  need
https://graphql-code-generator.com/docs/plugins/typescript#avoidoptionals-boolean-default-value-false

**Comment by david-golightly-leapyear** [comment]
@standy I see - `avoidOptionals` looks to implement the correct behavior. I'm confused why it's not `default: true`, and indeed why the behavior in this PR would ever be desirable? Unless you're mixing your server & client types inappropriately... but IMO that should be discouraged, it goes against the grain of GraphQL.

**Comment by standy** [comment]
@david-golightly-leapyear
That PR fixes inconsistent behavior (described above).

**Comment by david-golightly-leapyear** [comment]
@standy I don't understand how it's inconsistent. The server ("top-level") types and the client ("query") types are often going to be inconsistent in that manner, and that seems to be _correct_ behavior.

The server has to support any possible query, so it can't ever specify the correct type for all queries.
The client _does_ know what fields it's requesting, and is in a position to say "field X is non-optional, because I'm requesting it and it should appear in the GraphQL response".

Indeed, when you look at the behavior of a GraphQL response, any field requested by the client is returned either with a value or `null`, but cruciall _never_ with `undefined` or with the field missing.

That's because `undefined` (aka the ES2019 "optional type") is not a GraphQL type, and the ES2019 "optional type" is also _not the same thing_ as what is called "optional type" in the GraphQL spec.

**Comment by david-golightly-leapyear** [comment]
tl;dr at a minimum, if this PR is going to be preserved, it should also make `avoidOptionals` default: true, since this PR breaks existing correct usage (avoiding use of `undefined` in generated client types).

**Comment by rizkyario** [comment]
Screenshot 2020-04-24 at 17 43 26

Is this fix available in flow plugins?
seems doesn't work here, or am I missing some config?
